src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java

```diff
@@ -639,30 +632,30 @@ default ProgrammingExercise findByIdWithTemplateAndSolutionParticipationTeamAssi
     }
 
     /**
-     * Finds a programming exercise by its id including the submissions for its solution and template participation.
+     * Finds a {@link ProgrammingExercise} by its ID with optional dynamic fetching of associated entities.
      *
-     * @param exerciseId          The id of the programming exercise.
-     * @param withGradingCriteria True, if the grading instructions of the exercise should be included as well.
-     * @return A programming exercise that has the given id.
-     * @throws EntityNotFoundException In case no exercise with the given id exists.
+     * @param exerciseId   the ID of the programming exercise to find.
+     * @param fetchOptions a collection of {@link ProgrammingExerciseFetchOptions} indicating which associated entities to fetch.
+     * @return the {@link ProgrammingExercise} with the specified ID and the associated entities fetched according to the provided options.
+     * @throws EntityNotFoundException if the programming exercise with the specified ID does not exist.
      */
     @NotNull
-    default ProgrammingExercise findByIdWithAuxiliaryRepositoriesTeamAssignmentConfigAndGradingCriteriaElseThrow(long exerciseId, boolean withGradingCriteria)
-            throws EntityNotFoundException {
-
-        final Specification<ProgrammingExercise> specification = (root, query, criteriaBuilder) -> {
-            root.fetch(Exercise_.CATEGORIES, JoinType.LEFT);
-            root.fetch(Exercise_.TEAM_ASSIGNMENT_CONFIG, JoinType.LEFT);
-            root.fetch(ProgrammingExercise_.AUXILIARY_REPOSITORIES, JoinType.LEFT);
-
-            if (withGradingCriteria) {
-                root.fetch(Exercise_.GRADING_CRITERIA, JoinType.LEFT);
-            }
-
-            return null;
-        };
+    default ProgrammingExercise findByIdWithDynamicFetchElseThrow(long exerciseId, ProgrammingExerciseFetchOptions... fetchOptions) throws EntityNotFoundException {
+        return findByIdWithDynamicFetchElseThrow(exerciseId, Set.of(fetchOptions));
+    }
```

Comment:
This also looks like a nice ultiitly method that could be placed in `DynamicSpecificationRepository`.

Reply:
I currently do not see a good way to move this method, without also moving the other `findByIdWithDynamicFetchElseThrow`. The latter cannot be moved without reintroducing `RepositoryImpl`, which I'll discuss with Stephan first and would likely do in a separate PR, as it can also provide other benefits.

Reply:
I talked with Stephan and will do this in a follow-up where I reintroduce `RepositoryImpl`